core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -319,14 +323,26 @@ public ShardRouting activePrimary(ShardId shardId) {
     /**
      * Returns one active replica shard for the given shard id or <code>null</code> if
      * no active replica is found.
+     *
+     * Since replicas could possibly be on nodes with a newer version of ES than
+     * the primary is, this will return replicas on the lowest version of ES.
      */
     public ShardRouting activeReplica(ShardId shardId) {
+        Version lowestVersionSeen = null;
+        ShardRouting candidate = null;
         for (ShardRouting shardRouting : assignedShards(shardId)) {
             if (!shardRouting.primary() && shardRouting.active()) {
-                return shardRouting;
+                Version replicaNodeVersion = nodesToVersions.get(shardRouting.currentNodeId());
+                if (replicaNodeVersion == null && candidate == null) {
```

Comment:
Is it possible for `replicaNodeVersion` to be null (I don't think so)? Can you add an assertion stating != null instead of the if ( == null) check?